private/buf/cmd/buf/command/alpha/repo/reposync/reposync.go

```diff
@@ -0,0 +1,473 @@
+) error {
+	// TODO: resume from last sync point
+	return repo.ForEachCommit(branch, func(commit git.Commit) error {
+		for _, module := range modules {
+			if err := pushModuleCommit(
+				ctx,
+				container,
+				clientConfig,
+				module.dir,
+				module.identity,
+				repo,
+				branch,
+				commit,
+				storagegitProvider,
+				knownTags,
+				createWithVisibility,
+			); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func pushModuleCommit(
+	ctx context.Context,
+	container appflag.Container,
+	clientConfig *connectclient.Config,
+	moduleDir string,
+	moduleIdentityOverride bufmoduleref.ModuleIdentity,
+	repo git.Repository,
+	branch string,
+	commit git.Commit,
+	storagegitProvider storagegit.Provider,
+	knownTags map[string][]string,
+	createWithVisibility string,
+) error {
+	sourceBucket, err := storagegitProvider.NewReadBucket(
+		commit.Tree(),
+		storagegit.ReadBucketWithSymlinksIfSupported(),
+	)
+	if err != nil {
+		return err
+	}
+	sourceBucket = storage.MapReadBucket(sourceBucket, storage.MapOnPrefix(moduleDir))
+	foundModule, err := bufconfig.ExistingConfigFilePath(ctx, sourceBucket)
+	if err != nil {
+		return err
+	}
+	if foundModule == "" {
+		// We did not find a module. Carry on to the next commit.
+		return nil
+	}
+	sourceConfig, err := bufconfig.GetConfigForBucket(
+		ctx,
+		sourceBucket,
+	)
+	if err != nil {
+		// We found a module but the module config is invalid. We can warn on this
+		// and carry on. Note that because of resumption, we will typically only come
+		// across this commit once, we will not log this warning again.
+		container.Logger().Warn(
+			"invalid module at commit %s at path %s: %s\n",
+			zap.String("commit", commit.Hash().String()),
+			zap.String("module", moduleDir),
+			zap.Error(err),
+		)
+		return nil
+	}
+	if sourceConfig.ModuleIdentity == nil {
+		// Unnamed module. Carry on.
+		return nil
+	}
+	moduleIdentity := sourceConfig.ModuleIdentity
+	if moduleIdentityOverride != nil {
+		moduleIdentity = moduleIdentityOverride
+	}
+	builtModule, err := bufmodulebuild.BuildForBucket(
+		ctx,
+		sourceBucket,
+		sourceConfig.Build,
+	)
+	if err != nil {
+		// We failed to build the module. We can warn on this
+		// and carry on. Note that because of resumption, we will typically only come
+		// across this commit once, we will not log this warning again.
+		container.Logger().Warn(
+			"invalid module at commit %s at path %s: %s\n",
+			zap.String("commit", commit.Hash().String()),
+			zap.String("module", moduleDir),
+			zap.Error(err),
+		)
+		return nil
+	}
+	pin, err := pushOrCreate(
+		ctx,
+		clientConfig,
+		repo,
+		commit,
+		branch,
+		knownTags[commit.Hash().Hex()],
+		moduleIdentity,
+		builtModule,
+		createWithVisibility,
+	)
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeAlreadyExists {
+			// Module has identical content. The BSR has already created the relevant labels
+			// for us, so we can simply carry on.
+			return nil
+		}
```

Comment:
For this scenario, in case the latest push has the same content, but there may be new tags, should we still try to tag it?

Reply:
The BSR will handle that before it returns `AlreadyExists`. PR for that is open in core.